src/nvim/tui/input.c

```diff
@@ -320,6 +321,60 @@ static bool handle_forced_escape(TermInput *input)
   return false;
 }
 
+static bool handle_background_color(TermInput *input)
+{
+  size_t count = 0;
+  size_t component = 0;
+  uint16_t rgb[] = { 0, 0, 0 };
+  uint16_t rgb_max[] = { 0, 0, 0 };
+  bool eat_backslash = false;
+  bool done = false;
+  bool bad = false;
+  if (rbuffer_size(input->read_stream.buffer) >= 9
+      && !rbuffer_cmp(input->read_stream.buffer, "\x1b]11;rgb:", 9)) {
+    rbuffer_consumed(input->read_stream.buffer, 9);
+    RBUFFER_EACH(input->read_stream.buffer, c, i) {
+      count = i + 1;
+      if (eat_backslash) {
+        done = true;
+        break;
+      } else if (c == '\x07') {
+        done = true;
+        break;
+      } else if (c == '\x1b') {
+        eat_backslash = true;
+      } else if (bad) {
+        /* ignore */
+      } else if (c == '/') {
+        if (component < 3)
+          component++;
+      } else if (isxdigit(c)) {
+        if (component < 3 && rgb_max[component] != 0xffff) {
+          uint16_t digit = (uint16_t)((c >= '0' && c <= '9') ? c - '0' : 10 + tolower(c) - 'a');
+          rgb_max[component] = (uint16_t)((rgb_max[component] << 4) | 0xf);
+          rgb[component] = (uint16_t)((rgb[component] << 4) | digit);
+        }
+      } else {
+        bad = true;
+      }
+    }
+    rbuffer_consumed(input->read_stream.buffer, count);
+    if (done && !bad && rgb_max[0] && rgb_max[1] && rgb_max[2]) {
+      double r = (double)rgb[0] / (double)rgb_max[0];
+      double g = (double)rgb[1] / (double)rgb_max[1];
+      double b = (double)rgb[2] / (double)rgb_max[2];
+      double luminance = 0.299*r + 0.587*g + 0.114*b; /* CCIR 601 */
+      char *bgvalue = luminance < 0.5 ? "dark" : "light";
+      set_string_default("bg", bgvalue, false);
+      if (!option_was_set((char_u *)"bg")) {
+        set_option_value((char_u *)"bg", 0, (char_u *)bgvalue, 0);
+      }
+    }
```

Comment:
Try putting this block in a function e.g. `set_bg_color_deferred`, then schedule it on the main thread [like this](https://github.com/neovim/neovim/pull/5027/files#diff-9fe05c714b192fe85bb903baf3355196R401)

    loop_schedule(&main_loop, event_create(1, input_done_event, 0));